### DISCUSSION TITLE
M1 Support Confusion

### DISCUSSION THREAD

**Discussion by jwhitehorn** [Homebrew/discussions#2539] [Everyday usage]
I know there's a few other threads on this, but this didn't feel right in any one of them. Specifically, I'm confused what explicit steps someone must take to successfully utilize M1/Apple Silicon optimized formula.

I recently upgraded laptops, and used Time Machine to restore, so all my existing packages were moved over with me. From reading in [other threads](https://github.com/Homebrew/discussions/discussions/417) I see where it's recommended to uninstall brew (using the script), and then install it again. This, allegedly, would make use of the /opt path and install (if available) M1 optimized packages. However, this has not been my experience.

Specifically:

1. After migrating to my new M1 Mac w/ OS 12.0.1 (and installing Rosetta 2 for other purposes), I noticed that fish (my default shell) was Intel based.
2. From [the Formulae site](https://formulae.brew.sh/formula/fish#default), I see that Fish _should_ support native M1 with my OS/Hardware.
3. I uninstall all of Homebrew with the official script.
4. I re-install Homebrew with the official script.
5. I install fish (`brew install fish`)
6. I noticed that fish is NOT located in /opt nor is it an M1 binary.

So, what gives? Is there some official documentation for how one should "upgrade"? Should having uninstalled and re-installed not accomplished what I am expecting?

**Comment by gromgit** [comment]
What's the output of `brew config` and `brew doctor`?

**Reply by jwhitehorn** [reply]
`brew config` confirms that the fresh install is indeed running under Rosetta 2. `brew doctor` indicates that all is well.

**Reply by gromgit** [reply]
And what's the value of `HOMEBREW_PREFIX` in `brew config`?

_Sidenote_: When folks who are helping you debug an issue ask for command output, they generally aren't looking for a one-liner summary thereof.

**Reply by jwhitehorn** [reply]
I'm not sure if the sidenote is intended to be snarky, but I appreicate the thought. I cannot provide you the full output, since I've since uninstalled brew (again). I'm reinstalling a fresh from a shell that I know is running outside of Rosetta.

**Reply by gromgit** [reply]
No snark intended, just a world-weary statement of fact, from years of being led on wild goose chases by folks who insisted that I didn't need to know the gory details to help them fix their problems, only whatever they chose to tell me. They were, almost to a fault, completely and utterly wrong.

**Comment by carlocab** [comment]
Did you run the install script under Rosetta? That's likely what happened given what you're seeing. It's possible your terminal was running under Rosetta without you noticing it.

**Reply by gromgit** [reply]
I don't have an M1 box to test it, but would running the install script in a Rosetta shell with `arch -arm64` create an M1 Homebrew setup?

**Reply by jwhitehorn** [reply]
How does one know if the terminal is running under Rosetta? As a built-in app shouldn't it be M1 native? How does one exit out of that? Shouldn't the install script handle this itself?

Not to throw eggs, but there is a level of polish here that is missing. I'm also not sure why there isn't a more seamless upgrade path for existing users, short of uninstalling everything.

**Reply by carlocab** [reply]
> How does one know if the terminal is running under Rosetta? 

You can tell by running

```
sysctl -a sysctl.proc_translated
```

If I that shows a value of `1`, then you are running under Rosetta. `brew config` should also be able to tell you.

> As a built-in app shouldn't it be M1 native?

It should, yes. The only way to get it to run under Rosetta is if it were configured, perhaps inadvertently, to do so by the user.

> How does one exit out of that?

It depends on how exactly your terminal came to be running under Rosetta. But you can always make sure you're executing the install script natively by prepending `arch -arm64` to the `/bin/bash` command you use to install Homebrew.

> Shouldn't the install script handle this itself?

Well, the install script can't read your mind about whether you want a native installation of Homebrew in `/opt/homebrew` or a Rosetta one in `/usr/local`, so it can only go by whether you're running Rosetta when it's invoked. Pull requests implementing improvements to this are welcome.

> Not to throw eggs, but there is a level of polish here that is missing.

We're a volunteer-run project who work on Homebrew for free in our spare time and ask for nothing in return. Rough edges are inevitable. As above, pull requests to smooth out those edges are welcome as well.

> I'm also not sure why there isn't a more seamless upgrade path for existing users, short of uninstalling everything.

Package management is difficult. Other package managers even ask you to reinstall everything when you upgrade your OS, even if you're still using the same machine.

You can make the process a bit easier if you use `brew bundle`. You'll have seen how to use this in the other threads about migrating your Homebrew installation.

**Reply by jwhitehorn** [reply]
Thank you @carlocab. My questions were mostly rhetorical, as your original response helped me figure out what to do. FWIW it's not the terminal that was/can run under Rosetta, but the shell. As for telling if the shell is running under Rosetta, `uname -a` presents this as well.

As for intent, and "reading minds", I think it's safe to assume that anyone who _purchased_ a new Mac, would prefer anything that can run natively should. Also, in regards to superhuman abilities, it doesn't take much foresight to also see into the future and realize that Rosetta is a _temporary_ solution - therefore legacy solutions should be avoided for new installs.

I understand that this is a volunteer effort, but that does not mean that I cannot call a duck a duck. Every project has it's room for improvement. I'm just simply providing some feedback from the perspective of an end-user. 

Thanks for all your help!

**Reply by carlocab** [reply]
> As for telling if the shell is running under Rosetta, `uname -a` presents this as well.

This often works but is unreliable. For example, if you happen to have `uname` from a Rosetta installation of `coreutils` in your `PATH`, then `uname -a` will still show `x86_64`, regardless of whether you are running under Rosetta or not. Using `sysctl` is a much safer way of determining it.

> I think it's safe to assume that anyone who _purchased_ a new Mac, would prefer anything that can run natively should.

No, this is really not safe to assume. There is still a small but significant fraction of software in Homebrew that don't work natively on ARM. There are users who use these software, so there will inevitably be users who will want a Rosetta installation of Homebrew.

> Rosetta is a _temporary_ solution

This is true, but I'm not sure you appreciate how long "temporary" can be. For reference, the first Rosetta framework was around for about ~6 years. This incarnation of Rosetta is likely to stick around for at least that long, and it's almost certain that there will still be software in Homebrew that doesn't support M1 natively when Apple decides to stop shipping Rosetta. I suspect this version will be around for even longer than the last one, in fact. Back then, Apple was switching into the most widely used processor architecture for laptops/desktops. Now, Apple is doing the opposite. Rosetta installations of Homebrew, and the users who want them, are likely to be around for quite some time.

> I'm just simply providing some feedback from the perspective of an end-user.

Sure, but to be frank: your feedback is largely ill-informed and misguided. Most of it is based around optimising Homebrew (or, at least, Homebrew's installer) for your personal use cases, opinions, and mistaken beliefs. Homebrew has more than one user, and not all these users are the same.

This isn't to say I think the install script is perfect. It certainly has room for improvement, but your specific suggestions about where that room is are not correct.

**Reply by jwhitehorn** [reply]
> Sure, but to be frank: your feedback is largely ill-informed and misguided.

Lol, ok. I'm not going to be baited into a fight with a troll. I appreciate that you're not interested in improving this. Heck, even a PROMPT to the user like "hey, you're on a M1 Mac, do you wish to install the legacy version of Homebrew [y/N]?" would be a million times more user friendly than the current setup. But, yes, ill-formed and misguided - got it, lol

**Reply by carlocab** [reply]
A prompt would break scripted installs. This is a common use case of the installer. This suggestion is yet another example of trying to optimise Homebrew for your use case without consideration for others.

> But, yes, ill-formed and misguided

Just calling a duck a duck.

**Reply by melissawood123** [reply]
I agree 🚫
